utils/SnapPy/Snappy/ICONMeteorologyCalculator.py

```diff
@@ -22,17 +22,44 @@
 '''
 
 from datetime import datetime, timedelta
-from glob import iglob
-import math
+import netCDF4
+import numpy as np
 import os
 import subprocess
-import time
 
 from Snappy.Resources import Resources, MetModel
 import Snappy.MeteorologyCalculator
 
 class ICONMeteorologyCalculator(Snappy.MeteorologyCalculator.MeteorologyCalculator):
     '''Calculate dwd icon-meteorology'''
+    dir_template = "NRPA_LON{lon0}_LAT{lat0}_ICON"
+
+
+    @staticmethod
+    def get_valid_timesteps(filename):
+        """Extract the valid timesteps usable for SNAP from a icon-netcdf file.
+
+        :param filename: filename of the netcdf-file
+        :return: (timesteps, times) a tuple of step-numbers and datetime-values for the valid timesteps
+        """
+        time_var = 'time'
+        x_wind_var = 'x_wind_pl'
+        with netCDF4.Dataset(filename, 'r') as nc:
+            times = netCDF4.num2date(nc['time'][:], nc['time'].units, only_use_python_datetimes=True)
+            x_wind = nc['x_wind_pl']
+
+            valid_times = []
+            for i, dt in enumerate(times):
+                wind0 = np.ma.filled(x_wind[i,0,0,0], np.nan) # convert masked array to float
```

Comment:
I prefer working with nan's in the data rather than with masked arrays + possible nans, so that is my usual handling of all input-data from the netCDF4 library.